File: core/src/main/kotlin/com/piotrwalkusz/smartlaw/core/model/rule/Rule.kt
package com.piotrwalkusz.smartlaw.core.model.rule

import com.piotrwalkusz.smartlaw.core.model.common.Id
import com.piotrwalkusz.smartlaw.core.model.element.Element
import com.piotrwalkusz.smartlaw.core.model.meta.MetaArgument
import com.piotrwalkusz.smartlaw.core.model.template.StaticTemplate
import com.piotrwalkusz.smartlaw.core.model.template.Template

data class Rule(
        val id: Id,
        val name: String,
        val description: String? = null,
        val arguments: List<MetaArgument> = listOf(),
        val content: Template<String>,
        val elements: Template<List<Element>> = StaticTemplate(listOf()),
        val outputs: Map<String, Template<*>> = mapOf(),
        val interfaces: List<Id> = listOf()
)

File: frontend/src/model/Rule.ts
import Id, { IdUtils } from "./Id";
import { List, Map } from "immutable";
import MetaArgument, { decodeMetaArgument } from "./MetaArgument";
import Template, { decodeTemplate } from "./Template";
import { decodeList, decodeMap, decodeNullable, decodeString } from "../utils/Decoders";
import Element, { elementMeta } from "./Element";

export const decodeRule = (json: any): Rule => {
  return {
    id: IdUtils.decode(json.id),
    name: decodeString(json.name),
    description: decodeNullable(json.description, decodeString),
    arguments: decodeList(json.arguments, decodeMetaArgument),
    content: decodeTemplate(json.content, decodeString),
    elements: decodeTemplate(json.elements, (json) => decodeList(json, elementMeta.decodeOrException)),
    outputs: decodeMap(json.outputs, (json) => decodeTemplate(json, (json) => json)),
    interfaces: decodeList(json.interfaces, IdUtils.decode),
  };
};

export default interface Rule {
  id: Id;
  name: string;
  description: string | null;
  arguments: List<MetaArgument>;
  content: Template<string>;
  elements: Template<List<Element>>;
  outputs: Map<string, Template<any>>;
  interfaces: List<Id>;
}
